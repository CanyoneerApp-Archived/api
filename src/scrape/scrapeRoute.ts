import {Feature} from '@turf/helpers';
import jsdom from 'jsdom';
import {
  GeoJSONRouteV2,
  IndexRouteV2,
  PermitV2 as Permit,
  RouteV2,
  permit1to2,
} from '../types/RouteV2';
import cachedFetch, {md5} from './cachedFetch';
import parseAdditionalRisk from './parseAdditionalRisk';
import {parseDescription} from './parseDescription';
import parseDifficulty, {getTechnicalGrade, getWaterGrade} from './parseDifficulty';
import parseKML from './parseKML';
import parseMonths from './parseMonths';
import {parseRaps} from './parseRaps';
import parseSport from './parseSports';
import {mostReleventElement, parseTable} from './parseTable';
import parseTime from './parseTime';

export async function scrapeRoute(url: string): Promise<RouteV2 | undefined> {
  const text = await cachedFetch(url);
  if (!text) return undefined;

  const {
    window: {document},
  } = new jsdom.JSDOM(text, {url});

  // This generally works but not for linked bluugnome data
  const kml = await parseKML(document);

  const tableElementRowMap = parseTable(document.querySelector('.tablecanyon tbody'));
  const raps = parseRaps(tableElementRowMap['Raps']);
  const qualityPopSection = tableElementRowMap['Rating'];
  const quality =
    qualityPopSection.querySelectorAll('.starRate4')?.length ??
    0 + (qualityPopSection.querySelectorAll('.starRate2')?.length ?? 0) / 2;
  const months = parseMonths(tableElementRowMap['Best season']);
  const vehicle = tableElementRowMap['Vehicle']?.textContent?.trim().replace('Vehicle:', '');

  // Typically we just need the last element
  const tableElements: {[key: string]: Element | undefined | null} = {};
  for (const key in tableElementRowMap) {
    tableElements[key] = mostReleventElement(key, tableElementRowMap[key]);
  }
  const rating = tableElements['Difficulty']?.textContent?.trim() ?? '';

  if (!parseSport(rating, ['canyoneering']).includes('canyoneering')) {
    return undefined;
  }

  const difficulty = parseDifficulty(rating);

  const geojson =
    kml.geoJSON?.type === 'FeatureCollection'
      ? kml.geoJSON
      : kml.geoJSON
        ? {type: 'FeatureCollection', features: [kml.geoJSON]}
        : undefined;

  const index: IndexRouteV2 = {
    id: md5(url),
    name: document.querySelector('h1')?.textContent ?? 'Unknown',
    quality: quality,
    months: months,
    additionalRisk: parseAdditionalRisk(rating),
    vehicle: vehicle,
    shuttle: tableElements['Shuttle']?.textContent?.trim(),
    permits: permit1to2[tableElements['Red Tape']?.textContent?.trim() ?? ''] as Permit | undefined,
    technicalGrade: getTechnicalGrade[difficulty ?? ''],
    waterGrade: getWaterGrade[difficulty ?? ''],
    timeGrade: parseTime(rating),
    rappelCountMin: raps.countMin,
    rappelCountMax: raps.countMax,
    rappelLengthMax: raps.lengthMax,
  };

  return {
    ...index,
    latitude: parseFloat(tableElements['Location']?.textContent?.split(',')[0] ?? ''),
    longitude: parseFloat(tableElements['Location']?.textContent?.split(',')[1] ?? ''),
    url: url,
    description: await parseDescription(document),
    geojson: geojson && {
      type: 'FeatureCollection',
      features: geojson.features.map(
        (feature: Feature) =>
          ({
            ...feature,
            properties: {
              ...feature.properties,
              ...Object.fromEntries(
                Object.entries(index).map(([key, value]) => [`route.${key}`, value]),
              ),
            },
          }) as unknown as GeoJSONRouteV2,
      ),
    },
  };
}
