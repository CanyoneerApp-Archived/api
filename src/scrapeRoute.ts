import cachedFetch from './utils/cachedFetch';
import jsdom from 'jsdom';
import parseDifficulty from './utils/parseDifficulty';
import parseTime from './utils/parseTime';
import parseSport from './utils/parseSports';
import parseAdditionalRisk from './utils/parseAdditionalRisk';
import {parseDescription} from './utils/parseDescription';
import {parseRaps} from './utils/parseRaps';
import parseKML from './utils/parseKML';
import {parseTable, mostReleventElement} from './utils/parseTable';
import parseMonths from './utils/parseMonths';

export async function scrapeRoute(url: string): Promise<any> {
    const text = await cachedFetch(url);
    const {window: {document}} = new jsdom.JSDOM(text, {url});

    // This generally works but not for linked bluugnome data
    const kml = await parseKML(document);

    const tableElementRowMap = parseTable(document.querySelector('.tablecanyon tbody'));
    const raps = parseRaps(tableElementRowMap['Raps']);
    const qualityPopSection = tableElementRowMap['Rating']!;
    const quality = qualityPopSection.querySelectorAll('.starRate4')?.length ?? 0 + (qualityPopSection.querySelectorAll('.starRate2')?.length ?? 0) / 2;
    const months = parseMonths(tableElementRowMap['Best season']);
    const vehicle = tableElementRowMap['Vehicle']?.textContent?.trim().replace('Vehicle:', '');

    // Typically we just need the last element
    const tableElements = {};
    for (const key in tableElementRowMap) {
        tableElements[key] = mostReleventElement(key, tableElementRowMap[key]);
    }
    const rating = tableElements['Difficulty']?.textContent.trim() ?? '';

    // popularity is currently broken
    const popularity = tableElements['StarRank'] && parseInt(tableElements['StarRank'].querySelector('.starRate > span')!.textContent!.slice(2));
    return {
        URL: url,
        Name: document.querySelector('h1')!.textContent!,
        Quality: quality,
        Popularity: popularity,
        Latitude: tableElements['Location'] && parseFloat(tableElements['Location'].textContent!.split(',')[0]),
        Longitude: tableElements['Location'] && parseFloat(tableElements['Location'].textContent!.split(',')[1]),
        Months: months,
        Difficulty: parseDifficulty(rating),
        AdditionalRisk: parseAdditionalRisk(rating),
        Vehicle: vehicle,
        Shuttle: tableElements['Shuttle']?.textContent.trim(),
        Permits: tableElements['Red Tape']?.textContent.trim(),
        Sports: parseSport(rating, ['canyoneering']),
        Time: parseTime(rating),
        RappelCountMin: raps.countMin,
        RappelCountMax: raps.countMax,
        RappelLengthMax: raps.lengthMax,
        KMLURL: kml.url,
        HTMLDescription: await parseDescription(document),
        GeoJSON: kml.geoJSON
    };
}
