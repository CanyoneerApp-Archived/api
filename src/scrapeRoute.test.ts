import {scrapeRoute} from './scrapeRoute';

// This integration test alerts us if our scraper starts returning different data.
// This could happen because of:
//   - a bug in our code -> fix the bug
//   - a change to RopeWiki's HTML structure -> fix our scraper
//   - a legitimate change to one of these canyon's beta -> update the snapshot by running  `yarn test --updateSnapshot`

// These canyons are chosen because they have a lot of metadata and but are unlikely to be updated frequently

describe('scrapeRoute', () => {
  it('matches snapshot for Cerebus', async () => {
    expect(
      transform(await scrapeRoute('https://ropewiki.com/Cerberus_Canyon_(North_Fork)')),
    ).toMatchSnapshot();
  });

  it('matches snapshot for Behunin', async () => {
    expect(transform(await scrapeRoute('https://ropewiki.com/Behunin_Canyon'))).toMatchSnapshot();
  });
});

function transform({GeoJSON, HTMLDescription, ...rest}: Awaited<ReturnType<typeof scrapeRoute>>) {
  return {
    HasGeoJSON: !!GeoJSON,
    HasHTMLDescription: !!HTMLDescription,
    ...rest,
  };
}
