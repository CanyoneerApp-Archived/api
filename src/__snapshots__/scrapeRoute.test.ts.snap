// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrapeRoute matches snapshot for Behunin 1`] = `
{
  "AdditionalRisk": undefined,
  "Difficulty": "3b",
  "HasGeoJSON": true,
  "HasHTMLDescription": true,
  "KMLURL": "https://ropewiki.com/images/c/cb/Behunin_Canyon.kml",
  "Latitude": 37.2825,
  "Longitude": -112.9697,
  "Months": [
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ],
  "Name": "Behunin Canyon",
  "Permits": "Permit required",
  "Popularity": undefined,
  "Quality": 3,
  "RappelCountMax": 9,
  "RappelCountMin": 9,
  "RappelLengthMax": 170,
  "Shuttle": "None",
  "Sports": [
    "canyoneering",
  ],
  "Time": "III",
  "URL": "https://ropewiki.com/Behunin_Canyon",
  "Vehicle": "Passenger",
}
`;

exports[`scrapeRoute matches snapshot for Cerebus 1`] = `
{
  "AdditionalRisk": undefined,
  "Difficulty": "4a",
  "HasGeoJSON": true,
  "HasHTMLDescription": true,
  "KMLURL": "https://ropewiki.com/images/2/2f/Cerberus_Canyon_%28North_Fork%29.kml",
  "Latitude": 36.2135,
  "Longitude": -116.7292,
  "Months": [
    "January",
    "Feburary",
    "March",
    "April",
    "November",
    "December",
  ],
  "Name": "Cerberus Canyon (North Fork)",
  "Permits": "No permit required",
  "Popularity": undefined,
  "Quality": 4,
  "RappelCountMax": 27,
  "RappelCountMin": 27,
  "RappelLengthMax": 270,
  "Shuttle": "Required 1 hour",
  "Sports": [
    "canyoneering",
  ],
  "Time": "IV",
  "URL": "https://ropewiki.com/Cerberus_Canyon_(North_Fork)",
  "Vehicle": "Passenger",
}
`;
